### DISCUSSION TITLE
Synch timeout for task waiting semaphore

### DISCUSSION THREAD

**Discussion by orbitcoms** [Mair/esp32-course#106] [Course Qeustions]
Hi Mair,
I was looking again at examples of using the Binary semaphores in the course and have a question about the timeout while waiting to "GET" a semaphore.

Can you synch the delay so it starts at a specific point?

What I am trying to do is send a UART data packet to an external device and process received packet but the time frame for the received packet can vary from several ms to several hundred ms and if I don't receive the response within say 10 seconds of sending the UART data then I want to retry.  If the wait routine is not in synch with the UART TX then it could be timing out when it should be waiting. Hope this make sense

Or is this not possible and needs to implement a timer or something?

**Comment by Mair** [comment]
@orbitcoms 
I think I need a little more context to understand the requirements. if you set your timeout to 10 secs wouldn't that mean that you could check the result of the semaphore after 10 secs get and if its not OK retry and, if it was OK then continue?

**Comment by orbitcoms** [comment]
But then the timeout could happen any time and not after the uart sends?
ie. The Task waiting for uart_data_ready will be running in an endless loop waiting for uart data semaphore to be set. If I happen to have send a data packet out uart say 20ms before the 10 second timeout, then the task will resend the data before there was a chance to get the response to packet just sent.

So it needs to be like this
1. Send uart data packet
2. wait for response to that packet for up to 10 whole seconds 
3. if response received before timeout then all good, else resend packet because we obviously got no response from the packet.
